app/modules/file.c

```diff
@@ -214,18 +214,71 @@ static int file_open( lua_State* L )
 static int file_list( lua_State* L )
 {
   vfs_dir  *dir;
+  const char *pattern;
+  static const char *emptystr = "";
+  struct vfs_stat stat;
+  int pcres;
+  int res = 1;
+
+  pattern = luaL_optlstring(L, 1, emptystr, NULL);
+
+  dir = vfs_opendir("");
+  if (dir == NULL) {
+    return 0;
+  }
+
+  /* Allocate result table T */
+  lua_newtable( L );
+
+  /* If we're doing a match, push the Pattern and Function */
+  if (pattern != emptystr) {
+    lua_pushstring( L, pattern );
+    /* 
+     * We know that pattern is a string, and so the "match" method will always
+     * exist.  No need to check return value here
+     */
+    luaL_getmetafield( L, -1, "match" );
+  }
+
+  while (vfs_readdir(dir, &stat) == VFS_RES_OK) {
+    if (pattern != emptystr) {
+      /* Stack: T P F .  Duplicate F, push Name, duplicate P */
+      lua_pushvalue( L, -1 );
+      lua_pushstring( L, stat.name );
+      lua_pushvalue( L, -4 );
+      /* Call transforms stack from T P F F N P to T P F Result */
+      pcres = lua_pcall( L, 2, 1, 0 );
+      if (pcres != 0) {
+        /* Abort!  Stack is T P F Error; clear out and push nil under Error */
+        lua_pushnil( L );
+        lua_replace( L, -5 );
+        lua_remove( L, -2 );
+        lua_remove( L, -2 );
+        res = 2;
+        goto out;
+      }
 
-  if (dir = vfs_opendir("")) {
-    lua_newtable( L );
-    struct vfs_stat stat;
-    while (vfs_readdir(dir, &stat) == VFS_RES_OK) {
-      lua_pushinteger(L, stat.size);
-      lua_setfield(L, -2, stat.name);
+      /* Check and pop Result, restoring stack to T P F */
+      if (!lua_isnil( L, -1 )) {
+        lua_pushinteger( L, stat.size );
+        lua_setfield( L, -5, stat.name );
+      }
+      lua_pop( L, 1 );
+    } else {
+      /* Stack just holds result table at the top; add result */
+      lua_pushinteger( L, stat.size );
```

Comment:
these to push/setfield calls can be folded easily

Reply:
I debated doing that.  I can, if you like.